letsencrypt/acme/messages.py

```diff
@@ -0,0 +1,491 @@
+        return cls(session_id=jobj["sessionID"],
+                   nonce=jose.b64decode(jobj["nonce"]),
+                   responses=jobj["responses"],
+                   signature=other.Signature.from_json(
+                       jobj["signature"], validate=False),
+                   contact=jobj.get("contact", []))
+
+
+@Message.register  # pylint: disable=too-few-public-methods
+class Certificate(Message):
+    """ACME "certificate" message.
+
+    :ivar certificate: The certificate (:class:`M2Crypto.X509.X509`
+        wrapped in :class:`letsencrypt.acme.util.ComparableX509`).
+
+    :ivar list chain: Chain of certificates (:class:`M2Crypto.X509.X509` wrapped
+        in :class:`letsencrypt.acme.util.ComparableX509` ).
+
+    """
+    acme_type = "certificate"
+    schema = util.load_schema(acme_type)
+    __slots__ = ("certificate", "chain", "refresh")
+
+    def _fields_to_json(self):
+        fields = {"certificate": self._encode_cert(self.certificate)}
+        if self.chain:
+            fields["chain"] = [self._encode_cert(cert) for cert in self.chain]
+        if self.refresh is not None:
+            fields["refresh"] = self.refresh
+        return fields
+
+    @classmethod
+    def _decode_cert(cls, b64der):
+        return util.ComparableX509(M2Crypto.X509.load_cert_der_string(
+            jose.b64decode(b64der)))
+
+    @classmethod
+    def _encode_cert(cls, cert):
+        return jose.b64encode(cert.as_der())
+
+    @classmethod
+    def _from_valid_json(cls, jobj):
+        return cls(certificate=cls._decode_cert(jobj["certificate"]),
+                   chain=[cls._decode_cert(cert) for cert in
+                          jobj.get("chain", [])],
+                   refresh=jobj.get("refresh"))
+
+
+@Message.register
+class CertificateRequest(Message):
+    """ACME "certificateRequest" message.
+
+    :ivar csr: Certificate Signing Request (:class:`M2Crypto.X509.Request`
+        wrapped in :class:`letsencrypt.acme.util.ComparableX509`.
+    :ivar signature: Signature (:class:`letsencrypt.acme.other.Signature`).
+
+    """
+    acme_type = "certificateRequest"
+    schema = util.load_schema(acme_type)
+    __slots__ = ("csr", "signature")
+
+    @classmethod
+    def create(cls, key, sig_nonce=None, **kwargs):
+        """Create signed "certificateRequest".
+
+        :param key: Key used for signing.
+        :type key: :class:`Crypto.PublicKey.RSA`
+
+        :param str sig_nonce: Nonce used for signature. Useful for testing.
+
+        :returns: Signed "certificateRequest" ACME message.
+        :rtype: :class:`CertificateRequest`
+
+        """
+        return cls(signature=other.Signature.from_msg(
+            kwargs["csr"].as_der(), key, sig_nonce), **kwargs)
+
+    def verify(self):
+        """Verify signature.
+
+        :returns: True iff ``signature`` can be verified, False otherwise.
+        :rtype: bool
+
+        """
+        # TODO: must also check that the public key encoded in the JWK object
```

Comment:
This seems like an extremely important TODO item (more important than the AuthorizationRequest although the signature should remain the same in both), but this is really only necessary for a server implementation.  We should probably use a .. warning:: just to make it clear.  People might miss the significance of the TODO and "missing key verification bugs" are quite common